partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rex/ExprCallDynamic.kt

```diff
@@ -0,0 +1,68 @@
+package org.partiql.eval.internal.operator.rex
+
+import org.partiql.eval.internal.Record
+import org.partiql.eval.internal.operator.Operator
+import org.partiql.spi.function.PartiQLFunction
+import org.partiql.spi.function.PartiQLFunctionExperimental
+import org.partiql.value.PartiQLValue
+import org.partiql.value.PartiQLValueExperimental
+import org.partiql.value.PartiQLValueType
+import org.partiql.value.missingValue
+
+/**
+ * This represents Dynamic Dispatch.
+ *
+ * For the purposes of efficiency, this implementation aims to reduce any re-execution of compiled arguments. It
+ * does this by avoiding the compilation of [Candidate.fn] and [Candidate.coercions] into
+ * [ExprCallStatic]'s. By doing this, this implementation can evaluate ([eval]) the input [Record], execute and gather the
+ * arguments, and pass the [PartiQLValue]s directly to the [Candidate.eval].
+ */
+internal class ExprCallDynamic(
+    private val candidates: List<Candidate>,
+    private val args: Array<Operator.Expr>
+) : Operator.Expr {
+
+    @OptIn(PartiQLValueExperimental::class)
+    override fun eval(record: Record): PartiQLValue {
+        val actualArgs = args.map { it.eval(record) }.toTypedArray()
```

Comment:
say you entry point is `unionOf(int, decimal) + unionOf(int, decimal)`

You candidates looks like the following: 

```
├──Candidate[tag=Plan-d55fb882]
            │  └──Fn[signature=PLUS__DECIMAL_ARBITRARY_DECIMAL_ARBITRARY__DECIMAL_ARBITRARY, tag=Plan-288ba71]
            ├──Candidate[tag=Plan-e871d747]
            │  ├──Fn[signature=PLUS__DECIMAL_ARBITRARY_DECIMAL_ARBITRARY__DECIMAL_ARBITRARY, tag=Plan-fa4547c]
            │  └──Fn[signature=CAST_DECIMAL_ARBITRARY__INT32__DECIMAL_ARBITRARY, tag=Plan-381675d7]
            ├──Candidate[tag=Plan-52c5cf2b]
            │  ├──Fn[signature=PLUS__DECIMAL_ARBITRARY_DECIMAL_ARBITRARY__DECIMAL_ARBITRARY, tag=Plan-d7c0d7bb]
            │  └──Fn[signature=CAST_DECIMAL_ARBITRARY__INT32__DECIMAL_ARBITRARY, tag=Plan-c86b5dbe]
            └──Candidate[tag=Plan-dfc42937]
               └──Fn[signature=PLUS__INT32_INT32__INT32, tag=Plan-dac73796]
```

Seems to me that one you complete the evaluation of args, then you only need to search the right function to call by grouping the unique function signature.

Reply:
I see that the coerce function has not been placed in the arg. My bad.